File: retail/bot.go
package retail

import (
	"math"
	"math/rand"

	"economy/company"
	"economy/market"
	"economy/technicals"
)

// stopLoss represents one resting protective order a bot has placed
// against its own position in one specific company. Checked
// unconditionally every tick before any discretionary logic runs —
// this mechanical, non-negotiable check is what makes stop-losses the
// mechanism that produces genuine liquidity sweeps: enough bots
// independently placing stops at similar, predictable levels (see
// StopPlacementStyle) creates a cluster of resting forced-exit
// triggers that, once price reaches them, fire together and
// accelerate the move exactly as a real sweep does.
type stopLoss struct {
	active    bool
	triggerAt float64 // price level, above (for a short) or below (for a long) entry
	isLong    bool
}

// takeProfit represents a resting limit/profit target order.
// Locking in gains releases capital back into cash, allowing retail
// participants to recycle buying power into subsequent opportunities.
type takeProfit struct {
	active    bool
	triggerAt float64 // price level, above entry (for long) or below entry (for short)
	isLong    bool
}

// watchState holds everything specific to one company on a bot's
// watchlist: its own technicals aggregator (momentum on company A
// tells you nothing about company B, so each watched company needs
// independent bar history), its own cooldown counter, and its own
// stop-loss and take-profit target.
type watchState struct {
	agg               *technicals.Aggregator
	cooldownRemaining int
	stop              stopLoss
	target            takeProfit
	ticksHeld         int
}

// SimulatedRetailTrader is a behavioral, non-conviction-driven market
// participant: it does not compute a fair-value estimate the way
// HedgeFund/Bank do (even its FundamentalOnly variant reads
// company.ReportedValue, never TrueValue — see package company's
// documentation on why). Its size, information source, reaction
// shape, and stop discipline are all governed by its SkillTier,
// InformationStyle, and Archetype, generated at construction rather
// than hand-tuned per instance.
//
// A single bot trades a small watchlist of companies (1-5, typically),
// not the full universe — matching how real retail traders follow a
// personal handful of names rather than the whole market. See
// crowd.go's GenerateWatchlistPool for how watchlists are assigned
// with realistic, attention-skewed concentration.
type SimulatedRetailTrader struct {
	id      string
	account *market.Account

	watchlist map[string]*company.Company // symbol -> company, this bot's covered names
	state     map[string]*watchState      // symbol -> this bot's per-company tracking state

	tier      SkillTier
	style     InformationStyle
	archetype Archetype

	rng *rand.Rand
}

// NewSimulatedRetailTrader constructs one bot watching the given
// companies. Exported mainly so GenerateWatchlistPool (see crowd.go)
// has something to call repeatedly; direct construction is also fine
// for tests wanting precise control over a single bot's watchlist,
// tier, style, and archetype.
func NewSimulatedRetailTrader(
	id string,
	acct *market.Account,
	watchlist []*company.Company,
	tier SkillTier,
	style InformationStyle,
	archetype Archetype,
	rng *rand.Rand,
) *SimulatedRetailTrader {
	tp := tierProfiles[tier]

	companies := make(map[string]*company.Company, len(watchlist))
	states := make(map[string]*watchState, len(watchlist))
	for _, c := range watchlist {
		companies[c.Symbol] = c
		states[c.Symbol] = &watchState{agg: technicals.NewAggregator(tp.AggregatorWindow, 200)}
	}

	return &SimulatedRetailTrader{
		id:        id,
		account:   acct,
		watchlist: companies,
		state:     states,
		tier:      tier,
		style:     style,
		archetype: archetype,
		rng:       rng,
	}
}

func (b *SimulatedRetailTrader) ID() string               { return b.id }
func (b *SimulatedRetailTrader) Account() *market.Account { return b.account }

// Watches reports whether this bot's watchlist includes the given
// symbol — exposed mainly for tests and orchestrator bookkeeping
// (e.g. deciding which bots to poll for which company without calling
// NextOrders speculatively on every bot for every symbol).
func (b *SimulatedRetailTrader) Watches(symbol string) bool {
	_, ok := b.watchlist[symbol]
	return ok
}

// netInventory reads this bot's own signed position size for the
// given symbol — a small private equivalent of agent.NetInventory,
// deliberately not shared across the package boundary (see package
// doc: retail imports nothing from agent, and vice versa).
func (b *SimulatedRetailTrader) netInventory(symbol string) float64 {
	pos, ok := b.account.Positions[symbol]
	if !ok {
		return 0
	}
	if pos.Side == market.Short {
		return -pos.Quantity
	}
	return pos.Quantity
}

func absF(v float64) float64 {
	if v < 0 {
		return -v
	}
	return v
}

// checkStopLoss returns a forced market order if this bot holds a
// position in symbol and price has crossed that company's resting
// stop, unconditionally — no archetype/style discretion applies here,
// matching how a real stop-loss is a mechanical order, not a judgment
// call made fresh each time it's tested.
func (b *SimulatedRetailTrader) checkStopLoss(symbol string, currentPrice float64) *market.Order {
	ws := b.state[symbol]
	if ws == nil || !ws.stop.active {
		return nil
	}

	triggered := false
	if ws.stop.isLong && currentPrice <= ws.stop.triggerAt {
		triggered = true
	}
	if !ws.stop.isLong && currentPrice >= ws.stop.triggerAt {
		triggered = true
	}
	if !triggered {
		return nil
	}

	qty := absF(b.netInventory(symbol))
	if qty <= 0 {
		ws.stop.active = false
		return nil
	}

	side := market.Sell
	if !ws.stop.isLong {
		side = market.Buy // covering a short
	}

	ws.stop.active = false
	tp := tierProfiles[b.tier]
	ws.cooldownRemaining = tp.CooldownTicksMin + b.rng.Intn(tp.CooldownTicksMax-tp.CooldownTicksMin+1)

	return &market.Order{AgentID: b.id, Side: side, Quantity: qty, IsMarket: true}
}

// placeStop sets a fresh stop-loss for symbol after a new position is
// opened, at a level determined by this bot's tier-driven
// StopPlacementStyle.
func (b *SimulatedRetailTrader) placeStop(ws *watchState, entryPrice float64, isLong bool, bars []technicals.Bar) {
	tp := tierProfiles[b.tier]

	var distance float64
	switch tp.StopPlacement {
	case RoundNumberStop:
		// Nearest round number below (long) or above (short) entry —
		// the most predictable, most sweepable placement, matching
		// how real inexperienced retail commonly places stops.
		roundTo := roundNumberIncrement(entryPrice)
		if isLong {
			distance = entryPrice - (roundDown(entryPrice, roundTo))
		} else {
			distance = roundUp(entryPrice, roundTo) - entryPrice
		}
		if distance <= 0 {
			distance = entryPrice * 0.02 // fallback if already sitting on a round number
		}
	case SwingLevelStop:
		high, low, ok := technicals.SwingHighLow(bars, min(len(bars), 10))
		if ok && isLong && low < entryPrice {
			distance = entryPrice - low
		} else if ok && !isLong && high > entryPrice {
			distance = high - entryPrice
		} else {
			distance = entryPrice * 0.03
		}
	case BufferedStop:
		atr, ok := technicals.ATR(bars, min(len(bars)-1, 14))
		if ok && atr > 0 {
			distance = atr * 2.5 // wider, ATR-scaled buffer — less predictable, less sweepable
		} else {
			distance = entryPrice * 0.05
		}
	}

	if isLong {
		ws.stop = stopLoss{active: true, triggerAt: entryPrice - distance, isLong: true}
	} else {
		ws.stop = stopLoss{active: true, triggerAt: entryPrice + distance, isLong: false}
	}
}

// checkTakeProfit returns a market order closing a profitable position
// when price achieves this bot's profit target, locking in gains and
// releasing buying power back into cash.
func (b *SimulatedRetailTrader) checkTakeProfit(symbol string, currentPrice float64) *market.Order {
	ws := b.state[symbol]
	if ws == nil || !ws.target.active {
		return nil
	}

	triggered := false
	if ws.target.isLong && currentPrice >= ws.target.triggerAt {
		triggered = true
	}
	if !ws.target.isLong && currentPrice <= ws.target.triggerAt {
		triggered = true
	}
	if !triggered {
		return nil
	}

	qty := absF(b.netInventory(symbol))
	if qty <= 0 {
		ws.target.active = false
		return nil
	}

	side := market.Sell
	if !ws.target.isLong {
		side = market.Buy
	}

	ws.target.active = false
	if ws.stop.active {
		ws.stop.active = false
	}
	tp := tierProfiles[b.tier]
	ws.cooldownRemaining = tp.CooldownTicksMin + b.rng.Intn(tp.CooldownTicksMax-tp.CooldownTicksMin+1)

	return &market.Order{AgentID: b.id, Side: side, Quantity: qty, IsMarket: true}
}

// checkTimeDecayRotation rebalances or exits stagnant positions held for > 40 ticks
// so capital is not trapped in flat conditions, ensuring ongoing market turnover.
func (b *SimulatedRetailTrader) checkTimeDecayRotation(symbol string) *market.Order {
	ws := b.state[symbol]
	if ws == nil {
		return nil
	}

	inv := b.netInventory(symbol)
	if inv == 0 {
		ws.ticksHeld = 0
		return nil
	}

	ws.ticksHeld++
	if ws.ticksHeld < 40 {
		return nil
	}

	// 15% probability per tick to rotate stagnant capital once held > 40 ticks
	if b.rng.Float64() >= 0.15 {
		return nil
	}

	ws.ticksHeld = 0
	ws.stop.active = false
	ws.target.active = false

	qty := absF(inv)
	side := market.Sell
	if inv < 0 {
		side = market.Buy
	}

	tp := tierProfiles[b.tier]
	ws.cooldownRemaining = tp.CooldownTicksMin + b.rng.Intn(tp.CooldownTicksMax-tp.CooldownTicksMin+1)
	return &market.Order{AgentID: b.id, Side: side, Quantity: qty, IsMarket: true}
}

// placeTakeProfit sets a realistic profit target based on Archetype.
// When hit, profit is locked in, releasing cash so the bot can trade again.
func (b *SimulatedRetailTrader) placeTakeProfit(ws *watchState, entryPrice float64, isLong bool) {
	var pct float64
	switch b.archetype {
	case Disciplined:
		pct = 0.04 + b.rng.Float64()*0.04 // 4-8%
	case MomentumChaser:
		pct = 0.08 + b.rng.Float64()*0.10 // 8-18%
	case Contrarian:
		pct = 0.05 + b.rng.Float64()*0.06 // 5-11%
	case Degenerate:
		pct = 0.10 + b.rng.Float64()*0.15 // 10-25%
	default:
		pct = 0.06 + b.rng.Float64()*0.08 // 6-14%
	}

	var triggerPrice float64
	if isLong {
		triggerPrice = entryPrice * (1.0 + pct)
	} else {
		triggerPrice = entryPrice * (1.0 - pct)
		if triggerPrice <= 0.05 {
			triggerPrice = 0.05
		}
	}

	ws.target = takeProfit{
		active:    true,
		triggerAt: triggerPrice,
		isLong:    isLong,
	}
}

func roundNumberIncrement(price float64) float64 {
	switch {
	case price < 10:
		return 0.5
	case price < 100:
		return 5
	case price < 1000:
		return 50
	default:
		return 500
	}
}

func roundDown(price, increment float64) float64 {
	return float64(int(price/increment)) * increment
}

func roundUp(price, increment float64) float64 {
	down := roundDown(price, increment)
	if down == price {
		return price
	}
	return down + increment
}

// computeSignal builds this tick's Signal for one watched company
// from that company's own technicals window and (if this bot's
// style/tier allow) its fundamentals — see InformationStyle doc for
// what each style actually reads.
func (b *SimulatedRetailTrader) computeSignal(target *company.Company, ws *watchState, state market.MarketState) (Signal, bool) {
	bars := ws.agg.Bars()
	tp := tierProfiles[b.tier]

	var s Signal
	haveTechnical := false

	if momentum, ok := technicals.Momentum(bars, tp.MomentumWindow); ok {
		s.Momentum = momentum
		haveTechnical = true
	}

	if bb, ok := technicals.BollingerBands(bars, min(len(bars), 20), 2); ok && bb.Middle > 0 {
		s.Volatility = (bb.Upper - bb.Lower) / bb.Middle
	}

	haveFundamental := false
	if tp.ReactsToFundamentalEvents && state.HasMid && state.Mid > 0 {
		gap := (target.ReportedValue - state.Mid) / state.Mid
		s.FundamentalGap = gap
		haveFundamental = true
	}

	switch b.style {
	case TechnicalOnly:
		return s, haveTechnical
	case FundamentalOnly:
		if !haveFundamental {
			return s, false
		}
		// Repurpose Momentum as the drive signal for archetype
		// reaction functions, which are written in terms of Momentum/
		// Volatility — a fundamental gap plays the same directional
		// role technical momentum would for a TechnicalOnly bot.
		s.Momentum = s.FundamentalGap
		return s, true
	case Blended:
		if !haveTechnical && !haveFundamental {
			return s, false
		}
		if haveTechnical && haveFundamental {
			s.Momentum = (s.Momentum + s.FundamentalGap) / 2
		} else if haveFundamental {
			s.Momentum = s.FundamentalGap
		}
		return s, true
	case Confused:
		// Inconsistent by design: each tick, randomly pick which
		// signal (if any) actually drives this bot's reaction,
		// modeling a genuinely unstable process rather than a
		// diluted blend of both.
		roll := b.rng.Float64()
		switch {
		case roll < 0.4 && haveTechnical:
			return s, true
		case roll < 0.7 && haveFundamental:
			s.Momentum = s.FundamentalGap
			return s, true
		default:
			return s, false // this tick, nothing coherent drives it
		}
	}
	return s, false
}

// NextOrders implements market.OrderSource. Called once per tick per
// watched company by the orchestrator (mirroring how agent.Bank is
// driven — see Bank.NextOrders for the identical pattern). If
// state.Symbol isn't on this bot's watchlist, it has nothing to do
// this call, matching how a real retail trader simply doesn't react
// to a company they don't follow.
func (b *SimulatedRetailTrader) NextOrders(state market.MarketState) []*market.Order {
	target, watched := b.watchlist[state.Symbol]
	if !watched {
		return nil
	}
	ws := b.state[state.Symbol]

	if !state.HasMid || state.Mid <= 0 {
		return nil
	}
	ws.agg.AddTick(state.Mid)

	// Stop-loss check is unconditional and comes before cooldown or discretionary logic
	if order := b.checkStopLoss(state.Symbol, state.Mid); order != nil {
		return []*market.Order{order}
	}

	// Take-profit target check: locks in gains when target is reached, recycling cash
	if order := b.checkTakeProfit(state.Symbol, state.Mid); order != nil {
		return []*market.Order{order}
	}

	// Stagnant position rotation: frees up capital so market never freezes
	if order := b.checkTimeDecayRotation(state.Symbol); order != nil {
		return []*market.Order{order}
	}

	if ws.cooldownRemaining > 0 {
		ws.cooldownRemaining--
		return nil
	}

	signal, ok := b.computeSignal(target, ws, state)
	if !ok {
		return nil // not enough bar history yet, or (Confused) nothing driving this tick
	}

	tp := tierProfiles[b.tier]
	baseActivity := 0.15 // overall per-tick willingness to act at all, before archetype shaping
	probs := ReactionFor(b.archetype)(signal, baseActivity)

	roll := b.rng.Float64()
	var side market.Side
	switch {
	case roll < probs.Buy:
		side = market.Buy
	case roll < probs.Buy+probs.Sell:
		side = market.Sell
	default:
		return nil // hold
	}

	sizeFraction := tp.PositionSizeFractionMin + b.rng.Float64()*(tp.PositionSizeFractionMax-tp.PositionSizeFractionMin)
	inv := b.netInventory(state.Symbol)

	var quantity float64
	if side == market.Sell && inv > 0 {
		// Closing or trimming an existing long position
		quantity = math.Max(1.0, math.Floor(inv*sizeFraction))
		if quantity > inv {
			quantity = inv
		}
		if quantity <= 0 {
			return nil
		}
	} else if side == market.Buy && inv < 0 {
		// Covering or trimming an existing short position
		shortQty := -inv
		quantity = math.Max(1.0, math.Floor(shortQty*sizeFraction))
		if quantity > shortQty {
			quantity = shortQty
		}
		if quantity <= 0 {
			return nil
		}
	} else {
		// Opening a new position or adding to exposure using available buying power
		buyingPower := b.account.AvailableBuyingPower(map[string]float64{state.Symbol: state.Mid})
		notional := buyingPower * sizeFraction
		if notional <= 0 {
			return nil
		}
		quantity = math.Floor(notional / state.Mid)
		if quantity <= 0 {
			return nil
		}
		b.placeStop(ws, state.Mid, side == market.Buy, ws.agg.Bars())
		b.placeTakeProfit(ws, state.Mid, side == market.Buy)
	}

	ws.cooldownRemaining = tp.CooldownTicksMin + b.rng.Intn(tp.CooldownTicksMax-tp.CooldownTicksMin+1)
	return []*market.Order{{AgentID: b.id, Side: side, Quantity: quantity, IsMarket: true}}
}

File: src/retail/bot.rs
use std::collections::HashMap;
use std::sync::Arc;
use parking_lot::RwLock;
use rand::rngs::StdRng;
use rand::Rng;

use crate::company::Company;
use crate::market::{Account, MarketState, Order, OrderSource, Side};
use crate::technicals::{atr, bollinger_bands, momentum, swing_high_low, Aggregator, Bar};
use super::reaction::{reaction_for, Signal};
use super::taxonomy::{
    get_tier_profile, Archetype, InformationStyle, SkillTier, StopPlacementStyle,
};

#[derive(Debug, Clone, Copy, Default)]
pub struct StopLoss {
    pub active: bool,
    pub trigger_at: f64,
    pub is_long: bool,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct TakeProfit {
    pub active: bool,
    pub trigger_at: f64,
    pub is_long: bool,
}

#[derive(Debug, Clone)]
pub struct WatchState {
    pub agg: Aggregator,
    pub cooldown_remaining: usize,
    pub stop: StopLoss,
    pub target: TakeProfit,
    pub ticks_held: usize,
}

pub struct SimulatedRetailTrader {
    pub id: String,
    pub account: Arc<RwLock<Account>>,
    pub watchlist: HashMap<String, Company>,
    pub state: HashMap<String, WatchState>,
    pub tier: SkillTier,
    pub style: InformationStyle,
    pub archetype: Archetype,
    pub rng: StdRng,
}

impl SimulatedRetailTrader {
    pub fn new(
        id: impl Into<String>,
        account: Arc<RwLock<Account>>,
        watchlist: Vec<Company>,
        tier: SkillTier,
        style: InformationStyle,
        archetype: Archetype,
        rng: StdRng,
    ) -> Self {
        let tp = get_tier_profile(tier);
        let mut companies = HashMap::with_capacity(watchlist.len());
        let mut states = HashMap::with_capacity(watchlist.len());

        for c in watchlist {
            states.insert(
                c.symbol.clone(),
                WatchState {
                    agg: Aggregator::new(tp.aggregator_window, 200),
                    cooldown_remaining: 0,
                    stop: StopLoss::default(),
                    target: TakeProfit::default(),
                    ticks_held: 0,
                },
            );
            companies.insert(c.symbol.clone(), c);
        }

        Self {
            id: id.into(),
            account,
            watchlist: companies,
            state: states,
            tier,
            style,
            archetype,
            rng,
        }
    }

    pub fn watches(&self, symbol: &str) -> bool {
        self.watchlist.contains_key(symbol)
    }

    fn net_inventory(&self, symbol: &str) -> f64 {
        let acct = self.account.read();
        crate::agent::net_inventory(&acct, symbol)
    }

    fn check_stop_loss(&mut self, symbol: &str, current_price: f64) -> Option<Order> {
        let ws = self.state.get_mut(symbol)?;
        if !ws.stop.active {
            return None;
        }

        let mut triggered = false;
        if ws.stop.is_long && current_price <= ws.stop.trigger_at {
            triggered = true;
        }
        if !ws.stop.is_long && current_price >= ws.stop.trigger_at {
            triggered = true;
        }
        if !triggered {
            return None;
        }

        let is_long = ws.stop.is_long;
        let tp = get_tier_profile(self.tier);
        let cooldown = tp.cooldown_ticks_min
            + self.rng.gen_range(0..=(tp.cooldown_ticks_max - tp.cooldown_ticks_min));

        let qty = self.net_inventory(symbol).abs();
        let ws = self.state.get_mut(symbol)?;
        if qty <= 0.0 {
            ws.stop.active = false;
            return None;
        }

        let side = if is_long { Side::Sell } else { Side::Buy };
        ws.stop.active = false;
        ws.cooldown_remaining = cooldown;

        Some(Order {
            id: 0,
            agent_id: self.id.clone(),
            side,
            price: 0.0,
            quantity: qty,
            is_market: true,
        })
    }

    fn place_stop_for_ws(
        tier: SkillTier,
        ws: &mut WatchState,
        entry_price: f64,
        is_long: bool,
        bars: &[Bar],
    ) {
        let tp = get_tier_profile(tier);
        let distance = match tp.stop_placement {
            StopPlacementStyle::RoundNumberStop => {
                let round_to = round_number_increment(entry_price);
                let dist = if is_long {
                    entry_price - round_down(entry_price, round_to)
                } else {
                    round_up(entry_price, round_to) - entry_price
                };
                if dist <= 0.0 {
                    entry_price * 0.02
                } else {
                    dist
                }
            }
            StopPlacementStyle::SwingLevelStop => {
                let window = bars.len().min(10);
                if let Some((high, low)) = swing_high_low(bars, window) {
                    if is_long && low < entry_price {
                        entry_price - low
                    } else if !is_long && high > entry_price {
                        high - entry_price
                    } else {
                        entry_price * 0.03
                    }
                } else {
                    entry_price * 0.03
                }
            }
            StopPlacementStyle::BufferedStop => {
                let window = bars.len().saturating_sub(1).min(14);
                if let Some(a) = atr(bars, window) {
                    if a > 0.0 {
                        a * 2.5
                    } else {
                        entry_price * 0.05
                    }
                } else {
                    entry_price * 0.05
                }
            }
        };

        if is_long {
            ws.stop = StopLoss {
                active: true,
                trigger_at: entry_price - distance,
                is_long: true,
            };
        } else {
            ws.stop = StopLoss {
                active: true,
                trigger_at: entry_price + distance,
                is_long: false,
            };
        }
    }

    fn check_take_profit(&mut self, symbol: &str, current_price: f64) -> Option<Order> {
        let ws = self.state.get_mut(symbol)?;
        if !ws.target.active {
            return None;
        }

        let mut triggered = false;
        if ws.target.is_long && current_price >= ws.target.trigger_at {
            triggered = true;
        }
        if !ws.target.is_long && current_price <= ws.target.trigger_at {
            triggered = true;
        }
        if !triggered {
            return None;
        }

        let is_long = ws.target.is_long;
        let tp = get_tier_profile(self.tier);
        let cooldown = tp.cooldown_ticks_min
            + self.rng.gen_range(0..=(tp.cooldown_ticks_max - tp.cooldown_ticks_min));

        let qty = self.net_inventory(symbol).abs();
        let ws = self.state.get_mut(symbol)?;
        if qty <= 0.0 {
            ws.target.active = false;
            return None;
        }

        let side = if is_long { Side::Sell } else { Side::Buy };
        ws.target.active = false;
        ws.stop.active = false;
        ws.cooldown_remaining = cooldown;

        Some(Order {
            id: 0,
            agent_id: self.id.clone(),
            side,
            price: 0.0,
            quantity: qty,
            is_market: true,
        })
    }

    fn check_time_decay_rotation(&mut self, symbol: &str) -> Option<Order> {
        let inv = self.net_inventory(symbol);
        let ws = self.state.get_mut(symbol)?;
        if inv == 0.0 {
            ws.ticks_held = 0;
            return None;
        }

        ws.ticks_held += 1;
        if ws.ticks_held < 40 {
            return None;
        }

        if self.rng.gen::<f64>() >= 0.15 {
            return None;
        }

        ws.ticks_held = 0;
        ws.stop.active = false;
        ws.target.active = false;

        let qty = inv.abs();
        let side = if inv < 0.0 { Side::Buy } else { Side::Sell };

        let tp = get_tier_profile(self.tier);
        ws.cooldown_remaining = tp.cooldown_ticks_min
            + self.rng.gen_range(0..=(tp.cooldown_ticks_max - tp.cooldown_ticks_min));

        Some(Order {
            id: 0,
            agent_id: self.id.clone(),
            side,
            price: 0.0,
            quantity: qty,
            is_market: true,
        })
    }

    fn place_take_profit_for_ws(
        archetype: Archetype,
        rng: &mut StdRng,
        ws: &mut WatchState,
        entry_price: f64,
        is_long: bool,
    ) {
        let pct = match archetype {
            Archetype::Disciplined => 0.04 + rng.gen::<f64>() * 0.04,
            Archetype::MomentumChaser => 0.08 + rng.gen::<f64>() * 0.10,
            Archetype::Contrarian => 0.05 + rng.gen::<f64>() * 0.06,
            Archetype::PanicProne => 0.03 + rng.gen::<f64>() * 0.05,
            Archetype::Degenerate => 0.02 + rng.gen::<f64>() * 0.15,
        };

        if is_long {
            ws.target = TakeProfit {
                active: true,
                trigger_at: entry_price * (1.0 + pct),
                is_long: true,
            };
        } else {
            ws.target = TakeProfit {
                active: true,
                trigger_at: entry_price * (1.0 - pct),
                is_long: false,
            };
        }
    }

    fn compute_signal(
        style: InformationStyle,
        tier: SkillTier,
        rng: &mut StdRng,
        target: &Company,
        bars: &[Bar],
        state: &MarketState,
    ) -> Option<Signal> {
        let tp = get_tier_profile(tier);

        let mut s = Signal::default();
        let mut have_technical = false;

        if let Some(mom) = momentum(bars, tp.momentum_window) {
            s.momentum = mom;
            have_technical = true;
        }

        let bb_window = bars.len().min(20);
        if let Some(bb) = bollinger_bands(bars, bb_window, 2.0) {
            if bb.middle > 0.0 {
                s.volatility = (bb.upper - bb.lower) / bb.middle;
            }
        }

        let mut have_fundamental = false;
        if tp.reacts_to_fundamental_events {
            if let Some(mid) = state.mid {
                if mid > 0.0 {
                    s.fundamental_gap = (target.reported_value - mid) / mid;
                    have_fundamental = true;
                }
            }
        }

        match style {
            InformationStyle::TechnicalOnly => {
                if have_technical {
                    Some(s)
                } else {
                    None
                }
            }
            InformationStyle::FundamentalOnly => {
                if have_fundamental {
                    s.momentum = s.fundamental_gap;
                    Some(s)
                } else {
                    None
                }
            }
            InformationStyle::Blended => {
                if !have_technical && !have_fundamental {
                    return None;
                }
                if have_technical && have_fundamental {
                    s.momentum = (s.momentum + s.fundamental_gap) / 2.0;
                } else if have_fundamental {
                    s.momentum = s.fundamental_gap;
                }
                Some(s)
            }
            InformationStyle::Confused => {
                let roll = rng.gen::<f64>();
                if roll < 0.4 && have_technical {
                    Some(s)
                } else if roll < 0.7 && have_fundamental {
                    s.momentum = s.fundamental_gap;
                    Some(s)
                } else {
                    None
                }
            }
        }
    }
}

impl OrderSource for SimulatedRetailTrader {
    fn id(&self) -> &str {
        &self.id
    }

    fn next_orders(&mut self, state: &MarketState) -> Vec<Order> {
        let target = match self.watchlist.get(&state.symbol) {
            Some(t) => t.clone(),
            None => return Vec::new(),
        };

        let current_mid = match state.mid {
            Some(m) if m > 0.0 => m,
            _ => return Vec::new(),
        };

        if let Some(ws) = self.state.get_mut(&state.symbol) {
            ws.agg.add_tick(current_mid);
        }

        if let Some(order) = self.check_stop_loss(&state.symbol, current_mid) {
            return vec![order];
        }

        if let Some(order) = self.check_take_profit(&state.symbol, current_mid) {
            return vec![order];
        }

        if let Some(order) = self.check_time_decay_rotation(&state.symbol) {
            return vec![order];
        }

        let (bars, cooldown) = match self.state.get(&state.symbol) {
            Some(w) => (w.agg.all_bars(), w.cooldown_remaining),
            None => return Vec::new(),
        };

        if cooldown > 0 {
            if let Some(ws) = self.state.get_mut(&state.symbol) {
                ws.cooldown_remaining -= 1;
            }
            return Vec::new();
        }

        let signal = match Self::compute_signal(self.style, self.tier, &mut self.rng, &target, &bars, state) {
            Some(s) => s,
            None => return Vec::new(),
        };

        let tp = get_tier_profile(self.tier);
        let base_activity = 0.15;
        let probs = reaction_for(self.archetype, signal, base_activity);

        let roll = self.rng.gen::<f64>();
        let side = if roll < probs.buy {
            Side::Buy
        } else if roll < probs.buy + probs.sell {
            Side::Sell
        } else {
            return Vec::new(); // hold
        };

        let size_fraction = tp.position_size_fraction_min
            + self.rng.gen::<f64>() * (tp.position_size_fraction_max - tp.position_size_fraction_min);
        let inv = self.net_inventory(&state.symbol);

        let quantity: f64;

        if side == Side::Sell && inv > 0.0 {
            let q = (inv * size_fraction).floor().max(1.0);
            quantity = q.min(inv);
            if quantity <= 0.0 {
                return Vec::new();
            }
        } else if side == Side::Buy && inv < 0.0 {
            let short_qty = -inv;
            let q = (short_qty * size_fraction).floor().max(1.0);
            quantity = q.min(short_qty);
            if quantity <= 0.0 {
                return Vec::new();
            }
        } else {
            let buying_power = {
                let mut prices = HashMap::new();
                prices.insert(state.symbol.clone(), current_mid);
                let acct = self.account.read();
                acct.available_buying_power(&prices)
            };
            let notional = buying_power * size_fraction;
            if notional <= 0.0 {
                return Vec::new();
            }
            quantity = (notional / current_mid).floor();
            if quantity <= 0.0 {
                return Vec::new();
            }

            let tier = self.tier;
            let arch = self.archetype;
            if let Some(ws) = self.state.get_mut(&state.symbol) {
                Self::place_stop_for_ws(tier, ws, current_mid, side == Side::Buy, &bars);
                Self::place_take_profit_for_ws(arch, &mut self.rng, ws, current_mid, side == Side::Buy);
            }
        }

        if let Some(ws) = self.state.get_mut(&state.symbol) {
            ws.cooldown_remaining = tp.cooldown_ticks_min
                + self.rng.gen_range(0..=(tp.cooldown_ticks_max - tp.cooldown_ticks_min));
        }

        vec![Order {
            id: 0,
            agent_id: self.id.clone(),
            side,
            price: 0.0,
            quantity,
            is_market: true,
        }]
    }
}

fn round_number_increment(price: f64) -> f64 {
    if price < 10.0 {
        0.5
    } else if price < 100.0 {
        5.0
    } else if price < 1000.0 {
        50.0
    } else {
        500.0
    }
}

fn round_down(price: f64, increment: f64) -> f64 {
    ((price / increment).floor()) * increment
}

fn round_up(price: f64, increment: f64) -> f64 {
    let down = round_down(price, increment);
    if (down - price).abs() < 1e-9 {
        price
    } else {
        down + increment
    }
}
